onnx/defs/controlflow/defs.cc

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "onnx/defs/schema.h"
+using namespace onnx;
+
+using SupportType = onnx::OpSchema::SupportType;
+
+OPERATOR_SCHEMA(If)
+    .SetSupportLevel(SupportType::EXPERIMENTAL)
+    .SetDoc("If conditional")
+    .Input(0, "cond", "Condition for the if", "B")
+    .Output(
+        0,
+        "outputs",
+        "Values that are live-out to the enclosing scope.",
+        "V",
+        OpSchema::Variadic)
+    .Attr(
+        "then_branch",
+        "Graph to run if condition is true. Has N outputs: values you wish to "
+        "be live-out to the enclosing scope. The number of outputs must match"
+        " the number of outputs in the else_branch.",
+        AttributeProto::GRAPH,
+        true)
+    .Attr(
+        "else_branch",
+        "Graph to run if condition is false. Has N outputs: values you wish to"
+        " be live-out to the enclosing scope. The number of outputs must match"
+        " the number of outputs in the then_branch.",
+        AttributeProto::GRAPH,
+        true)
+    .TypeConstraint("V", OpSchema::all_tensor_types, "All Tensor types")
+    .TypeConstraint("B", {"tensor(bool)"}, "Only bool");
+
+OPERATOR_SCHEMA(Loop)
+    .SetSupportLevel(SupportType::EXPERIMENTAL)
+    .SetDoc(R"DOC(
+Generic Looping construct. This loop has multiple termination conditions:
+
```

Comment:
I personally feel that it will be better to have simpler primitives (like a simple while-loop) that can be composed to get this effect instead of having a single complex operator that combines multiple things (a for-loop and a while-loop together).

Reply:
We take the philosophy of having higher-level operators, so that backends don't have to engage in complex pattern-matching algorithms to identify the semantics of the network and reason about which transformations are legal. Could you give an example of what primitives you would propose instead?

Reply:
(a) The function-proposal is one way to avoid this philosophical tradeoff between (1) A smaller set of core primitives and (2) A long list of higher-level operators. A common higher-level operator can be defined as a function, easily recognized and optimally implemented by the backend. 

(b) Are there realistic examples that combine trip-count with a while-condition? From an efficiency and higher-level operator perspective, I would think that a "scan" operator (like CNTK's "Recurrence"), without any trip-count, would be a common pattern worth capturing as an operator/function.

I am not saying that we shouldn't define a complex operator like this if there is a compelling motivation. But it is unclear to me whether the current motivation is to define the basic/core control-flow primitives or define a higher-level operator for a specific class of situations. If needed, we could do both.

(c) The primitive I would propose is a while-loop without any trip-counts (and/or a "scan" operator)

Reply:
(a) Let me discuss this with the team. Will update later. But my instinctual response is that control flow operators are sufficiently changing the core semantics of the spec (in terms of analysis, optimization, etc) that it's okay to special-case this

(b) Beam search with a dynamic stopping condition as well as a max trip count is an example. This is a concrete use case we have that we want to support in ONNX

(c) The trip count gives semantic information to backends for optimization purposes. In the case where we have a trip count, backends can bulk allocate all required memory and/or do more fancy things like emit specific kernels for a given sequence length, etc. This would be harder to do if we have to pattern match the indexing machinery